### ISSUE TITLE
Page number Links not working correctly. Next and Previous not working as expected.

### ISSUE THREAD

**Issue by mayordwells** [michaelbromley/ngx-pagination#206]
**Angular version**: 4.4.5

**ngx-pagination version**: 3.0.1

**Description of issue**: I made a request to the server by using some ruby gems in a rails application. 
I used the following gems: 
[gem 'api-pagination'](https://github.com/davidcelis/api-pagination)
[gem 'kaminari'](https://github.com/kaminari/kaminari)
I followed the set up instruction as per the documentation for ngx-pagination.
What is really happening now is that I can't completely make use of the features available in the plugin. Here's what my current working code looks like: 
And sample demo of how it works can be seen here - [https://deaninfotech.herokuapp.com/dashboard/ng/search](https://deaninfotech.herokuapp.com/dashboard/ng/search)
In my component file I have this: 
```
page: number = 1;
  // loading: boolean;
total: number;

getProducts(page: number) {
    // this.loading = true;
    // const perPage = 20;
    // const start = (page - 1) * perPage;
    // const end = start + perPage;

    // The total number of items is more than 50,000 but I made it 1,000 for now because I still can't 
    // figure out how to get the total from the server. I'm leaving this issue out for later.
    this.total = 1000;
    this.productService.getProducts(page)
      .subscribe(
        products => this.products = products,
        error => this.errorMessage = <any>error
      );
  }
``` 
In ngOnInit I have:
```
ngOnInit(): void {
    this.getProducts(this.page);
}
```
This is how I got the products from the server in the service file:
```
private productsUrl = 'https://deaninfotech.herokuapp.com/products';

getProducts(page?: Number): Observable<Product[]> {
    return this.http.get(`${this.productsUrl}?page=${page}`)
                .map((response: Response) => <Product[]>response.json()
                  .filter(<Product>(product) => product.images != null && product.name != null))
                .catch(this.handleError);
  }
```
In the template.html file I have this: 
```
<ng-container *ngFor="let product of products | paginate: { id: 'server', itemsPerPage: 20, currentPage: page, totalItems: total }">
</ng-container>

<pagination-controls
      (pageChange)="getProducts($event)" 
      id="server"
    >
    </pagination-controls>
```
And this is what my server code in Rails look like:
```
def product_data
    product_data = paginate Product.all, per_page: 20
    render json: JSON.pretty_generate(product_data.as_json)
  end

  def index
    product_data
  end
```
Check here to see the JSON objects as well - [http://deaninfotech.herokuapp.com/products](http://deaninfotech.herokuapp.com/products.json)
Paginating the objects from server works like so, to see objects on page 50 for example [http://deaninfotech.herokuapp.com/products?page=50](http://deaninfotech.herokuapp.com/products?page=50)

**Steps to reproduce**: If you click on the pagination links,
 playing around with it you will see that it's definitely not giving the fully expected results,
 please guide me on what to do fix this, check the link here:
 [https://deaninfotech.herokuapp.com/dashboard/ng/search](https://deaninfotech.herokuapp.com/dashboard/ng/search)

**Expected result**: Previous, Next and Page Numbers between 5 and 50 should be clickable isn't it? When I click on Page 5, the other pages maybe between 5 and 10 should be available for clicking.

**Actual result**: the Previous Link not working, other numbered links not working as expected. 

**Demo**: [https://deaninfotech.herokuapp.com/dashboard/ng/search](https://deaninfotech.herokuapp.com/dashboard/ng/search)

If the code and explanation provided above won't be enough I can push the whole code on Github for you to see. 

Thanks in advance for your help. God bless you.

**Comment by michaelbromley** [comment]
Hi,

I can see two issues here:

1) You don't seem to be updating the value of `page`. Therefore, the value is always 1, even when clicking links. You could add something like this:

```TypeScript
getProducts(page?: Number): Observable<Product[]> {
    return this.http.get(`${this.productsUrl}?page=${page}`)
                .map((response: Response) => <Product[]>response.json()
                .filter(<Product>(product) => product.images != null && product.name != null))
                .do(() => this.page = page)
                .catch(this.handleError);
  }
```

2) Your API does not seem to return the total number of items, which you are already aware of. Looking at the API response, I can see that the total is actually returned as a header:

```
HTTP/1.1 304 Not Modified
Server: Cowboy
Date: Thu, 19 Oct 2017 19:14:38 GMT
Content-Length: 0
Connection: keep-alive
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
P3p: CP="Not used"
X-Frame-Options: ALLOWALL
Link: <http://deaninfotech.herokuapp.com/products.json?page=2357>; rel="last", <http://deaninfotech.herokuapp.com/products.json?page=2>; rel="next"
Per-Page: 20
Total: 47126  <----------------------------------------------------- right here
Etag: W/"de632949d25c97ded049e538dbc752d8"
Cache-Control: max-age=0, private, must-revalidate
X-Request-Id: 5839994f-cabe-4268-9f49-51d3ba7c6f53
X-Runtime: 0.022804
Via: 1.1 vegur
```
So you need to read that header and pass that value to your component's `total` variable.

Hope that helps!

**Comment by mayordwells** [comment]
Hi @michaelbromley 
Thanks for your answer, it was quite helpful, I have updated the service method accordingly. And it works.
```
getProducts(page?: Number): Observable<Product[]> {
    return this.http.get(`${this.productsUrl}?page=${page}`)
                .map((response: Response) => <Product[]>response.json()
                  .filter(<Product>(product) => product.images != null && product.images.AP != null && product.name != null && product.sku.AP != null ))
                .do(() => page = page)
                .catch(this.handleError);
  }
```
Now when you click on the page numbers it responds accordingly, check here: 
[https://deaninfotech.herokuapp.com/dashboard/ng/search](https://deaninfotech.herokuapp.com/dashboard/ng/search)

But the Previous and Next links are still not working as expected. And even the styling for the current page is still not updated to signify on page number the current page. 

From my Observation, I noticed that when you click on Next it does go to Page 2 and no more.
I couldn't do: ` .do(() => this.page = page) ` because the page variable was only declared in the component. 
So I changed it to this in the service method: 
`.do(() => page = page)`
The component remains the same as explained in my question above. 

Also, thanks for pointing me out in the right direction as regards the total. 
However, I noticed that because I am only filtering out the returned JSON on the client side, most of the pages don't display the complete 20 items per page. 

I am also still trying to figure out a way to group the items with the same SKU on the server side before paginating and sending to the client side and before getting the total. 

Anyway, how can we get the Next, Previous and 'Current Page' styling to work?

**Comment by michaelbromley** [comment]
> I couldn't do: .do(() => this.page = page) because the page variable was only declared in the component.
So I changed it to this in the service method:
.do(() => page = page)
The component remains the same as explained in my question above.

Is that API call taking place in a service? If so, then yeah, the `.do()` operator is pointless there. You just need to make sure that when the API call is successful, you set the `page` variable *in the component* to the new page value. Once the component `page` is updated, all the styling will work as expected.

> However, I noticed that because I am only filtering out the returned JSON on the client side, most of the pages don't display the complete 20 items per page.

What kind of filtering are you performing? With server-side paging, the usual thing would be to perform any kind of filtering on the backend. Otherwise you run into the exact problem that you mention.

**Comment by mayordwells** [comment]
Hi @michaelbromley,
You were right. I updated the page value in the component by adding ` this.page = page;`, I also removed the `.do()` call from the service, now everything works perfectly well. 
My component method now looks like this:
```
getProducts(page: number) {
    this.page = page;
    this.total = 2000; // remember to get the total from server response header
    this.productService.getProducts(page)
      .subscribe(
        products => this.products = products,
        error => this.errorMessage = <any>error
      );
  }
``` 
This is the line that performs the filtering on the client side in the service.
`.filter(<Product>(product) => product.images != null && product.images.AP != null && product.name != null && product.sku.AP != null ))`
Thanks for your answer on the filter as well, I'll work on it.
